In [1]:
import turicreate

In [2]:
people_data = turicreate.SFrame('people_wiki.sframe')
people_data.head()

Columns:
	URI	str
	name	str
	text	str

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+
|              text             |
+-------------------------------+
| digby morrell born 10 octo... |
| alfred j lewy aka sandy le... |
| harpdog brown is a singer ... |
| franz rottensteiner born i... |
| henry krvits born 30 decem... |
| sam henderson born october... |
| aaron lacrate is an americ... |
| trevor ferguson aka john f... |
| grant nelson born 27 april... |
| cathy caruth born 1955 is ... |
+-------------------------------+
[10 rows x 3 columns]

In [3]:
people_data.shape

(59071, 3)

# Exploring the dataset

In [5]:
obama = people_data[people_data['name'] == 'Barack Obama']
obama

Columns:
	URI	str
	name	str
	text	str

Rows: Unknown

Data:
+-------------------------------+--------------+-------------------------------+
|              URI              |     name     |              text             |
+-------------------------------+--------------+-------------------------------+
| <http://dbpedia.org/resour... | Barack Obama | barack hussein obama ii br... |
+-------------------------------+--------------+-------------------------------+
[? rows x 3 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [6]:
obama['text']

dtype: str
Rows: ?
['barack hussein obama ii brk husen bm born august 4 1961 is the 44th and current president of the united states and the first african american to hold the office born in honolulu hawaii obama is a graduate of columbia university and harvard law school where he served as president of the harvard law review he was a community organizer in chicago before earning his law degree he worked as a civil rights attorney and taught constitutional law at the university of chicago law school from 1992 to 2004 he served three terms representing the 13th district in the illinois senate from 1997 to 2004 running unsuccessfully for the united states house of representatives in 2000in 2004 obama received national attention during his campaign to represent illinois in the united states senate with his victory in the march democratic party primary his keynote address at the democratic national convention in july and his election to the senate in november he began his presidential campa

In [7]:
obama['word_count'] = turicreate.text_analytics.count_words(obama['text'])
obama

Columns:
	URI	str
	name	str
	text	str
	word_count	dict

Rows: 1

Data:
+-------------------------------+--------------+-------------------------------+
|              URI              |     name     |              text             |
+-------------------------------+--------------+-------------------------------+
| <http://dbpedia.org/resour... | Barack Obama | barack hussein obama ii br... |
+-------------------------------+--------------+-------------------------------+
+-------------------------------+
|           word_count          |
+-------------------------------+
| {'normalize': 1.0, 'sought... |
+-------------------------------+
[1 rows x 4 columns]

In [9]:
obama['word_count']

dtype: dict
Rows: 1
[{'normalize': 1.0, 'sought': 1.0, 'combat': 1.0, 'continued': 1.0, 'unconstitutional': 1.0, '8': 1.0, 'californias': 1.0, '1996': 1.0, 'marriage': 1.0, 'defense': 1.0, 'down': 1.0, 'proposition': 1.0, 'court': 1.0, 'supreme': 1.0, 'urged': 1.0, 'which': 1.0, 'briefs': 1.0, 'administration': 1.0, 'while': 1.0, 'americans': 1.0, 'called': 1.0, 'cuba': 1.0, 'gun': 1.0, 'related': 1.0, 'policies': 1.0, 'promoted': 1.0, '2013': 1.0, 'second': 2.0, 'romney': 1.0, 'filed': 1.0, '2012': 1.0, 'reelected': 1.0, 'taxpayer': 1.0, 'budget': 1.0, 'nations': 1.0, 'raise': 1.0, 'spending': 1.0, 'over': 1.0, 'lengthy': 1.0, 'gains': 1.0, 'seats': 1.0, '63': 1.0, 'total': 1.0, 'lost': 1.0, 'regained': 1.0, 'whether': 1.0, 'close': 1.0, 'patient': 1.0, 'by': 1.0, 'sandy': 1.0, 'after': 4.0, 'presidential': 2.0, 'november': 2.0, 'obama': 9.0, 'election': 3.0, 'august': 1.0, 'mccain': 1.0, 'primary': 2.0, 'he': 7.0, 'united': 3.0, 'with': 3.0, 'current': 1.0, 'campaign': 3.0, 'degree':

In [14]:
word_count_dict = obama['word_count']
word_count_dict

dtype: dict
Rows: 1
[{'normalize': 1.0, 'sought': 1.0, 'combat': 1.0, 'continued': 1.0, 'unconstitutional': 1.0, '8': 1.0, 'californias': 1.0, '1996': 1.0, 'marriage': 1.0, 'defense': 1.0, 'down': 1.0, 'proposition': 1.0, 'court': 1.0, 'supreme': 1.0, 'urged': 1.0, 'which': 1.0, 'briefs': 1.0, 'administration': 1.0, 'while': 1.0, 'americans': 1.0, 'called': 1.0, 'cuba': 1.0, 'gun': 1.0, 'related': 1.0, 'policies': 1.0, 'promoted': 1.0, '2013': 1.0, 'second': 2.0, 'romney': 1.0, 'filed': 1.0, '2012': 1.0, 'reelected': 1.0, 'taxpayer': 1.0, 'budget': 1.0, 'nations': 1.0, 'raise': 1.0, 'spending': 1.0, 'over': 1.0, 'lengthy': 1.0, 'gains': 1.0, 'seats': 1.0, '63': 1.0, 'total': 1.0, 'lost': 1.0, 'regained': 1.0, 'whether': 1.0, 'close': 1.0, 'patient': 1.0, 'by': 1.0, 'sandy': 1.0, 'after': 4.0, 'presidential': 2.0, 'november': 2.0, 'obama': 9.0, 'election': 3.0, 'august': 1.0, 'mccain': 1.0, 'primary': 2.0, 'he': 7.0, 'united': 3.0, 'with': 3.0, 'current': 1.0, 'campaign': 3.0, 'degree':

## The most important word is the,in,and which are not of use.

In [23]:
word_count_dict.stack(['words','count']).sort('count',ascending=False)

Columns:
	words	str
	count	float

Rows: 273

Data:
+-------+-------+
| words | count |
+-------+-------+
|  the  |  40.0 |
|   in  |  30.0 |
|  and  |  21.0 |
|   of  |  18.0 |
|   to  |  14.0 |
|  his  |  11.0 |
| obama |  9.0  |
|  act  |  8.0  |
|   a   |  7.0  |
|   he  |  7.0  |
+-------+-------+
[273 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

# TFIDF

In [26]:
people_data.head()

Columns:
	URI	str
	name	str
	text	str

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+
|              text             |
+-------------------------------+
| digby morrell born 10 octo... |
| alfred j lewy aka sandy le... |
| harpdog brown is a singer ... |
| franz rottensteiner born i... |
| henry krvits born 30 decem... |
| sam henderson born october... |
| aaron lacrate is an americ... |
| trevor ferguson aka john f... |
| grant nelson born 27 april... |
| cathy caruth born 1955 is ... |
+-------------------------------+
[10 rows x 3 columns]

In [27]:
people_data['word_count'] = turicreate.text_analytics.count_words(people_data['text'])
people_data['tfidf'] = turicreate.text_analytics.tf_idf(people_data['text'])
people_data.head()

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| digby morrell born 10 octo... | {'melbourne': 1.0, 'parade... |
| alfred j lewy aka sandy le... | {'time': 1.0, 'each': 1.0,... |
| harpdog brown is a singer ... | {'society': 1.0, 'hamilton... |
| franz rottensteiner born i... | {'kurdlawitzpreis': 1.0, '... |
| henry krvits born 30 decem... | {'curtis': 1.0, '2007': 1.... |
| sam henderson born october... | {'asses': 1.0, 'sic': 1.0,... |
| aaron lacrate is an americ... | {'streamz': 1.0, 'includin... |
| trevor ferguson aka john f... | {'concordia': 1.0, 'creati... |
| grant nelson born 27 april... | {'heavies': 1.0, 'new': 1.... |
| cathy caruth born 1955 is ... | {'2002': 1.0, 'harvard': 1... |
+-------------------------------+-------------------------------+
+-------------------------------+
|             tfidf             |
+-------------------------------+
| {'melbourne': 3.8914310119... |
| {'time': 1.325334207420049... |
| {'society': 2.444804726208... |
| {'kurdlawitzpreis': 10.986... |
| {'curtis': 5.2995200328853... |
| {'asses': 9.60020102810530... |
| {'streamz': 10.98649538922... |
| {'concordia': 6.2502969408... |
| {'heavies': 8.907053847545... |
| {'2002': 1.875312588782230... |
+-------------------------------+
[10 rows x 5 columns]

## Now Let's Check for Obama

In [31]:
obama.remove_column('tfidf',inplace=True)

Columns:
	URI	str
	name	str
	text	str
	word_count	dict

Rows: 1

Data:
+-------------------------------+--------------+-------------------------------+
|              URI              |     name     |              text             |
+-------------------------------+--------------+-------------------------------+
| <http://dbpedia.org/resour... | Barack Obama | barack hussein obama ii br... |
+-------------------------------+--------------+-------------------------------+
+-------------------------------+
|           word_count          |
+-------------------------------+
| {'normalize': 1.0, 'sought... |
+-------------------------------+
[1 rows x 4 columns]

In [33]:
obama_tfidf = people_data[people_data['name'] == 'Barack Obama']['tfidf']

### Now the Most Important Word is Obama. 

In [35]:
obama_tfidf.stack(['word','count']).sort('count',ascending=False)

Columns:
	word	str
	count	float

Rows: 273

Data:
+-------------+--------------------+
|     word    |       count        |
+-------------+--------------------+
|    obama    |  43.2956530720749  |
|     act     | 27.67822262297991  |
|     iraq    | 17.747378587965535 |
|   control   | 14.887060845181308 |
|     law     | 14.722935761763422 |
|   ordered   | 14.533373950913514 |
|   military  | 13.115932778499415 |
| involvement | 12.784385241175055 |
|   response  | 12.784385241175055 |
|  democratic | 12.410688697332166 |
+-------------+--------------------+
[273 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

# Document Retrieval

In [36]:
people_data.head()

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: 10

Data:
+-------------------------------+---------------------+
|              URI              |         name        |
+-------------------------------+---------------------+
| <http://dbpedia.org/resour... |    Digby Morrell    |
| <http://dbpedia.org/resour... |    Alfred J. Lewy   |
| <http://dbpedia.org/resour... |    Harpdog Brown    |
| <http://dbpedia.org/resour... | Franz Rottensteiner |
| <http://dbpedia.org/resour... |        G-Enka       |
| <http://dbpedia.org/resour... |    Sam Henderson    |
| <http://dbpedia.org/resour... |    Aaron LaCrate    |
| <http://dbpedia.org/resour... |   Trevor Ferguson   |
| <http://dbpedia.org/resour... |     Grant Nelson    |
| <http://dbpedia.org/resour... |     Cathy Caruth    |
+-------------------------------+---------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| digby morrell born 10 octo... | {'melbourne': 1.0, 'parade... |
| alfred j lewy aka sandy le... | {'time': 1.0, 'each': 1.0,... |
| harpdog brown is a singer ... | {'society': 1.0, 'hamilton... |
| franz rottensteiner born i... | {'kurdlawitzpreis': 1.0, '... |
| henry krvits born 30 decem... | {'curtis': 1.0, '2007': 1.... |
| sam henderson born october... | {'asses': 1.0, 'sic': 1.0,... |
| aaron lacrate is an americ... | {'streamz': 1.0, 'includin... |
| trevor ferguson aka john f... | {'concordia': 1.0, 'creati... |
| grant nelson born 27 april... | {'heavies': 1.0, 'new': 1.... |
| cathy caruth born 1955 is ... | {'2002': 1.0, 'harvard': 1... |
+-------------------------------+-------------------------------+
+-------------------------------+
|             tfidf             |
+-------------------------------+
| {'melbourne': 3.8914310119... |
| {'time': 1.325334207420049... |
| {'society': 2.444804726208... |
| {'kurdlawitzpreis': 10.986... |
| {'curtis': 5.2995200328853... |
| {'asses': 9.60020102810530... |
| {'streamz': 10.98649538922... |
| {'concordia': 6.2502969408... |
| {'heavies': 8.907053847545... |
| {'2002': 1.875312588782230... |
+-------------------------------+
[10 rows x 5 columns]

In [38]:
knn_model = turicreate.nearest_neighbors.create(people_data,features=['tfidf'],label='name')

Starting brute force nearest neighbors model training.

Validating distance components.

Initializing model data.

Initializing distances.

Done.

# Now Let's Check for Similarity

In [39]:
clinton = people_data[people_data['name'] == 'Bill Clinton']
clinton

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: Unknown

Data:
+-------------------------------+--------------+-------------------------------+
|              URI              |     name     |              text             |
+-------------------------------+--------------+-------------------------------+
| <http://dbpedia.org/resour... | Bill Clinton | william jefferson bill cli... |
+-------------------------------+--------------+-------------------------------+
+-------------------------------+-------------------------------+
|           word_count          |             tfidf             |
+-------------------------------+-------------------------------+
| {'presidents': 1.0, 'polls... | {'presidents': 4.520350664... |
+-------------------------------+-------------------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [44]:
knn_model.query(clinton)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 9.916ms      |

| Done         |         | 100         | 298.156ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+------------------------+--------------------+------+
| query_label |    reference_label     |      distance      | rank |
+-------------+------------------------+--------------------+------+
|      0      |      Bill Clinton      |        0.0         |  1   |
|      0      | Hillary Rodham Clinton | 0.7931034482758621 |  2   |
|      0      |     George W. Bush     | 0.7944325481798715 |  3   |
|      0      |      Dick Cheney       | 0.8064516129032258 |  4   |
|      0      |      Barack Obama      | 0.8138528138528138 |  5   |
+-------------+------------------------+--------------------+------+
[5 rows x 4 columns]

## Let's read about these people to know whether they are related or not?
We see that these all are politicians so they are related with each other

In [46]:
bush = people_data[people_data['name'] == 'George W. Bush']
bush['text']

dtype: str
Rows: ?
['george walker bush born july 6 1946 is an american politician and businessman who served as the 43rd president of the united states from 2001 to 2009 and the 46th governor of texas from 1995 to 2000 the eldest son of barbara and george h w bush he was born in new haven connecticut after graduating from yale university in 1968 and harvard business school in 1975 bush worked in oil businesses he married laura welch in 1977 and ran unsuccessfully for the house of representatives shortly thereafter he later coowned the texas rangers baseball team before defeating ann richards in the 1994 texas gubernatorial election bush was elected president in 2000 after a close and controversial election becoming the fourth president to be elected while receiving fewer popular votes nationwide than his opponent bush is the second president to have been the son of a former president the first having been john quincy adams he is also the brother of jeb bush former governor of floridae

In [47]:
jolie = people_data[people_data['name'] == 'Angelina Jolie']
jolie['text']

dtype: str
Rows: ?
['angelina jolie doli johlee born angelina jolie voight june 4 1975 is an american actress director producer writer and humanitarian she has won an academy award two screen actors guild awards and three golden globe awards and has been cited as hollywoods highestpaid actress jolie made her screen debut as a child alongside her father jon voight in lookin to get out 1982 her film career began in earnest a decade later with the lowbudget production cyborg 2 1993 followed by her first leading role in a major film hackers 1995 she starred in the critically acclaimed biographical television films george wallace 1997 and gia 1998 and won an academy award for best supporting actress for her performance in the drama girl interrupted 1999 jolies starring role as the video game heroine lara croft in lara croft tomb raider 2001 established her as a leading hollywood actress she continued her successful actionstar career with mr mrs smith 2005 wanted 2008 and salt 2010 and recei

In [48]:
knn_model.query(jolie)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 16.186ms     |

| Done         |         | 100         | 297.009ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+--------------------+--------------------+------+
| query_label |  reference_label   |      distance      | rank |
+-------------+--------------------+--------------------+------+
|      0      |   Angelina Jolie   |        0.0         |  1   |
|      0      |     Brad Pitt      | 0.7840236686390533 |  2   |
|      0      |   Julianne Moore   | 0.7958579881656804 |  3   |
|      0      | Billy Bob Thornton |  0.80306905370844  |  4   |
|      0      |   George Clooney   |     0.8046875      |  5   |
+-------------+--------------------+--------------------+------+
[5 rows x 4 columns]

# Assignment

## 1.

In [50]:
john = people_data[people_data['name'] ==  'Elton John']
john['text']

dtype: str
Rows: ?
['sir elton hercules john cbe born reginald kenneth dwight 25 march 1947 is an english singer songwriter composer pianist record producer and occasional actor he has worked with lyricist bernie taupin as his songwriter partner since 1967 they have collaborated on more than 30 albums to datein his fivedecade career elton john has sold more than 300 million records making him one of the bestselling music artists in the world he has more than fifty top 40 hits including seven consecutive no 1 us albums 58 billboard top 40 singles 27 top 10 four no 2 and nine no 1 for 31 consecutive years 19702000 he had at least one song in the billboard hot 100 his single something about the way you look tonightcandle in the wind 1997 sold over 33 million copies worldwide and is the bestselling single of all time he has received six grammy awards five brit awards winning two awards for outstanding contribution to music and the first brits icon in 2013 for his lasting impact on british 

In [51]:
john

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: Unknown

Data:
+-------------------------------+------------+-------------------------------+
|              URI              |    name    |              text             |
+-------------------------------+------------+-------------------------------+
| <http://dbpedia.org/resour... | Elton John | sir elton hercules john cb... |
+-------------------------------+------------+-------------------------------+
+-------------------------------+-------------------------------+
|           word_count          |             tfidf             |
+-------------------------------+-------------------------------+
| {'movements': 1.0, 'social... | {'movements': 5.0306580197... |
+-------------------------------+-------------------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [53]:
john_count_word = john['word_count']
john_count_word.stack(['word','count']).sort('count',ascending=False)

Columns:
	word	str
	count	float

Rows: 255

Data:
+-------+-------+
|  word | count |
+-------+-------+
|  the  |  27.0 |
|   in  |  18.0 |
|  and  |  15.0 |
|   of  |  13.0 |
|   a   |  10.0 |
|  has  |  9.0  |
|   he  |  7.0  |
|  john |  7.0  |
|   on  |  6.0  |
| award |  5.0  |
+-------+-------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

In [54]:
john_tfidf = john['tfidf']
john_tfidf.stack(['word','count']).sort('count',ascending=False)

Columns:
	word	str
	count	float

Rows: 255

Data:
+---------------+--------------------+
|      word     |       count        |
+---------------+--------------------+
|    furnish    | 18.38947183999428  |
|     elton     | 17.482320270031995 |
|   billboard   | 17.30368095754203  |
|      john     | 13.93931279239831  |
|  songwriters  | 11.250406447031539 |
|  overallelton | 10.986495389225194 |
| tonightcandle | 10.986495389225194 |
|   fivedecade  | 10.293348208665249 |
|    19702000   | 10.293348208665249 |
|      aids     | 10.262846934045534 |
+---------------+--------------------+
[255 rows x 2 columns]
Note: Only the head of the SFrame is printed.
You can use print_rows(num_rows=m, num_columns=n) to print more rows and columns.

### Highest Word Counts: the,in,and
### Highest tfidf Word Counts: furnish,elton,billboard

## 2.

In [55]:
v_b = people_data[people_data['name'] == 'Victoria Beckham']
v_b

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: Unknown

Data:
+-------------------------------+------------------+
|              URI              |       name       |
+-------------------------------+------------------+
| <http://dbpedia.org/resour... | Victoria Beckham |
+-------------------------------+------------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| victoria caroline beckham ... | {'week': 1.0, 'york': 1.0,... |
+-------------------------------+-------------------------------+
+-------------------------------+
|             tfidf             |
+-------------------------------+
| {'week': 3.57252509903475,... |
+-------------------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [56]:
p_m = people_data[people_data['name'] == 'Paul McCartney']
p_m

Columns:
	URI	str
	name	str
	text	str
	word_count	dict
	tfidf	dict

Rows: Unknown

Data:
+-------------------------------+----------------+
|              URI              |      name      |
+-------------------------------+----------------+
| <http://dbpedia.org/resour... | Paul McCartney |
+-------------------------------+----------------+
+-------------------------------+-------------------------------+
|              text             |           word_count          |
+-------------------------------+-------------------------------+
| sir james paul mccartney m... | {'children': 1.0, 'five': ... |
+-------------------------------+-------------------------------+
+-------------------------------+
|             tfidf             |
+-------------------------------+
| {'children': 2.42528531239... |
+-------------------------------+
[? rows x 5 columns]
Note: Only the head of the SFrame is printed. This SFrame is lazily evaluated.
You can use sf.materialize() to force materialization.

In [58]:
turicreate.distances.cosine(john['tfidf'][0],v_b['tfidf'][0])

0.9567006376655429

### distance between the articles on ‘Elton John’ and ‘Victoria Beckham’? ANS: 0.9567006376655429

In [59]:
turicreate.distances.cosine(john['tfidf'][0],p_m['tfidf'][0])

0.8250310029221779

### distance between the articles on ‘Elton John’ and Paul McCartney’? ANS: 0.8250310029221779

## 3.

In [60]:
knn_model_w_c = turicreate.nearest_neighbors.create(people_data,features=['word_count'],label='name',distance='cosine')

Starting brute force nearest neighbors model training.

Validating distance components.

Initializing model data.

Initializing distances.

Done.

In [61]:
knn_model_tfidf = turicreate.nearest_neighbors.create(people_data,features=['tfidf'],label='name',distance='cosine')

Starting brute force nearest neighbors model training.

Validating distance components.

Initializing model data.

Initializing distances.

Done.

In [62]:
knn_model_w_c.query(john)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 9.835ms      |

| Done         |         | 100         | 276.054ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+-------------------+-----------------------+------+
| query_label |  reference_label  |        distance       | rank |
+-------------+-------------------+-----------------------+------+
|      0      |     Elton John    | 2.220446049250313e-16 |  1   |
|      0      |   Cliff Richard   |  0.16142415258967036  |  2   |
|      0      |   Sandro Petrone  |  0.16822542751041114  |  3   |
|      0      |    Rod Stewart    |  0.16832716558706107  |  4   |
|      0      | Malachi O'Doherty |   0.177315545978884   |  5   |
+-------------+-------------------+-----------------------+------+
[5 rows x 4 columns]

In [63]:
knn_model_tfidf.query(john)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 12.88ms      |

| Done         |         | 100         | 300.659ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+------------------+------------------------+------+
| query_label | reference_label  |        distance        | rank |
+-------------+------------------+------------------------+------+
|      0      |    Elton John    | -2.220446049250313e-16 |  1   |
|      0      |   Rod Stewart    |   0.7172196678927374   |  2   |
|      0      |  George Michael  |   0.7476009989692848   |  3   |
|      0      | Sting (musician) |   0.7476719544306141   |  4   |
|      0      |   Phil Collins   |   0.7511932487904706   |  5   |
+-------------+------------------+------------------------+------+
[5 rows x 4 columns]

In [64]:
knn_model_w_c.query(v_b)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 6.415ms      |

| Done         |         | 100         | 257.837ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+--------------------------+------------------------+------+
| query_label |     reference_label      |        distance        | rank |
+-------------+--------------------------+------------------------+------+
|      0      |     Victoria Beckham     | -2.220446049250313e-16 |  1   |
|      0      | Mary Fitzgerald (artist) |  0.20730703611504997   |  2   |
|      0      |      Adrienne Corri      |  0.21450978278754795   |  3   |
|      0      |     Beverly Jane Fry     |  0.21746646874079278   |  4   |
|      0      |      Raman Mundair       |  0.21769547499150488   |  5   |
+-------------+--------------------------+------------------------+------+
[5 rows x 4 columns]

In [65]:
knn_model_tfidf.query(v_b)

Starting pairwise querying.

+--------------+---------+-------------+--------------+

| Query points | # Pairs | % Complete. | Elapsed Time |

+--------------+---------+-------------+--------------+

| 0            | 1       | 0.00169288  | 8.026ms      |

| Done         |         | 100         | 296.958ms    |

+--------------+---------+-------------+--------------+

Columns:
	query_label	int
	reference_label	str
	distance	float
	rank	int

Rows: 5

Data:
+-------------+---------------------+------------------------+------+
| query_label |   reference_label   |        distance        | rank |
+-------------+---------------------+------------------------+------+
|      0      |   Victoria Beckham  | 1.1102230246251565e-16 |  1   |
|      0      |    David Beckham    |   0.5481696102632145   |  2   |
|      0      | Stephen Dow Beckham |   0.7849867068283364   |  3   |
|      0      |        Mel B        |   0.8095855234085036   |  4   |
|      0      |    Caroline Rush    |    0.81982642291868    |  5   |
+-------------+---------------------+------------------------+------+
[5 rows x 4 columns]

* What’s the most similar article, other than itself, to the one on ‘Elton John’ using word count features?
### ANS: Cliff Richard
* What’s the most similar article, other than itself, to the one on ‘Elton John’ using TF-IDF features?
### ANS: Rod Stewart
* What’s the most similar article, other than itself, to the one on ‘Victoria Beckham’ using word count features?
### ANS: Mary Fitzgerald (artist)
* What’s the most similar article, other than itself, to the one on ‘Victoria Beckham’ using TF-IDF features?
### ANS: David Beckham